src/witan/send/adroddiad/ncy.clj

```diff
@@ -1,7 +1,298 @@
-(ns witan.send.adroddiad.ncy)
+(ns witan.send.adroddiad.ncy
+  "Definitions and functions for handling National Curriculum Year"
+  (:require [clojure.set :as set]
+            [tablecloth.api :as tc]
+            [hyperfiddle.rcf :as rcf]))
 
-(defn ncy->age [ncy]
-  (+ ncy 5))
 
-(defn age->ncy [age]
-  (- age 5))
+(defn- inclusive-range
+  "Returns a lazy seq of nums from `start` (inclusive) to `end` (inclusive), by step 1"
+  [start end]
+  (range start (inc end)))
+
+
+(def ncys
+  "Set of national curriculum years (NCY), coded numerically, with
+  reception as NCY 0 and earlier NCYs as negative integers."
+  (into (sorted-set) (inclusive-range -4 20)))
+
+
+(def ncy->age-at-start-of-school-year 
+  "Maps national curriculum year (NCY) to age in whole number of years
+  on 31st August prior to starting the school/academic year.
+
+  Early years NCYs are coded numerically, with reception as year 0 and
+  earlier NCYs as negative integers.
+
+  NCYs and ages for reception to NCY 11 (children aged 4 to 15 at the
+  start of the school/academic year) are per https://www.gov.uk/national-curriculum.
+  Extension to NCYs -4 to -1 (ages 0-3) and NCYs 12 to 20, (ages 16-24)
+  is by linear extrapolation, such that the offset between NCY and age
+  at the beginning of the school/academic year is +4.
+
+  Age at the start of the school/academic year is the age on 31st August
+  prior to the school/academic year. This is per:
+
+  - The [gov.uk website](https://www.gov.uk/schools-admissions/school-starting-age),
+  which (as of 23-NOV-2022) states:
+  \"Most children start school full-time in the September after their
+  fourth birthday. This means they’ll turn 5 during their first school
+  year. For example, if your child’s fourth birthday is between 1
+  September 2021 and 31 August 2022 they will usually start school in
+  September 2022.\".
+
+  - The [2022 SEN2 guide](https://assets.publishing.service.gov.uk/government/uploads/system/uploads/attachment_data/file/1013751/SEN2_2022_Guide.pdf),
+  which states in section 2, part 1, item 1.1 that: \"age breakdown
+  refers to age as at 31 August 2021\", implying (since this is for
+  the January 2022 SEN2 return) that age for SEN2 breakdowns is as of
+  31-AUG prior to starting the school year.
+
+  The maximum age for SEND is 25, but per the relevant legistlation
+  \"a local authority may continue to maintain an EHC plan for a young
+  person until the end of the academic year during which the young
+  person attains the age of 25\", such that they will have been 24 at
+  the start of that school/academic year, hence the maximum NCY in
+  this map is 20 (age 24 at start of school/academic year).
+  "
+  (apply sorted-map (interleave ncys (map #(+ % 4) ncys))))
```

Comment:
interesting approach. I would have probably written it as

```clojure
(into (sorted-map) 
  (map (fn [ncy] [(+ 4 ncy] ncy]))
 ncys)
```

but yours does the same thing, my version is just for interest

Reply:
I did have something like that at one point but stumbled across the `(apply sorted-map (interleave <keys> <vals>))` online and liked it as a) it separates the keys and vals vectors clearly (your suggestion has them the wrong way round for `ncy->age…` and I didn't spot that straight away), and b) makes the sorted map directly rather than via an into.

Is there a danger in using `(apply sorted-map (interleave <keys> <vals>))`? I ask because `zipmap` has a more sophisticated implementation.

So I guess another option that meets (a) and is likely more recognisable would be:
```
(into (sorted-map)
        (zipmap ncys (map #(+ % 4) ncys)))
```
?

Reply:
Per discussion, `(apply sorted-map (interleave <keys> <vals>))` approach is fine and can be left as is as:
- we should know `interleave`
- number of key-value pairs is small here so no issues with `apply` and a large seq.